File: packages/py-moose-lib/moose_lib/dmv2/materialized_view.py
"""
Materialized View definitions for Moose Data Model v2 (dmv2).

This module provides classes for defining Materialized Views,
including their SQL statements, target tables, and dependencies.
"""
from typing import Any, Optional, Union, Generic, TypeVar
from pydantic import BaseModel, ConfigDict, model_validator

from moose_lib import ClickHouseEngines
from ..utilities.sql import quote_identifier
from .types import BaseTypedResource, T
from .olap_table import OlapTable, OlapConfig
from .sql_resource import SqlResource


class MaterializedViewOptions(BaseModel):
    """Configuration options for creating a Materialized View.

    Attributes:
        select_statement: The SQL SELECT statement defining the view's data.
        select_tables: List of source tables/views the select statement reads from.
        table_name: (Deprecated in favor of target_table) Optional name of the underlying
                    target table storing the materialized data.
        materialized_view_name: The name of the MATERIALIZED VIEW object itself.
        engine: Optional ClickHouse engine for the target table (used when creating
                a target table via table_name or inline config).
        order_by_fields: Optional ordering key for the target table (required for
                         engines like ReplacingMergeTree).
        model_config: ConfigDict for Pydantic validation
    """
    select_statement: str
    select_tables: list[Union[OlapTable, SqlResource]]
    # Backward-compatibility: allow specifying just the table_name and engine
    table_name: Optional[str] = None
    materialized_view_name: str
    engine: Optional[ClickHouseEngines] = None
    order_by_fields: Optional[list[str]] = None
    metadata: Optional[dict] = None
    # Ensure arbitrary types are allowed for Pydantic validation
    model_config = ConfigDict(arbitrary_types_allowed=True)


class MaterializedView(SqlResource, BaseTypedResource, Generic[T]):
    """Represents a ClickHouse Materialized View.

    Encapsulates the MATERIALIZED VIEW definition and the underlying target `OlapTable`
    that stores the data.

    Args:
        options: Configuration defining the select statement, names, and dependencies.
        t: The Pydantic model defining the schema of the target table
           (passed via `MaterializedView[MyModel](...)`).

    Attributes:
        target_table (OlapTable[T]): The `OlapTable` instance storing the materialized data.
        config (MaterializedViewOptions): The configuration options used to create the view.
        name (str): The name of the MATERIALIZED VIEW object.
        model_type (type[T]): The Pydantic model associated with the target table.
        setup (list[str]): SQL commands to create the view and populate the target table.
        teardown (list[str]): SQL command to drop the view.
        pulls_data_from (list[SqlObject]): Source tables/views.
        pushes_data_to (list[SqlObject]): The target table.
    """
    target_table: OlapTable[T]
    config: MaterializedViewOptions

    def __init__(
            self,
            options: MaterializedViewOptions,
            target_table: Optional[OlapTable[T]] = None,
            **kwargs
    ):
        self._set_type(options.materialized_view_name, self._get_type(kwargs))

        # Resolve target table from options
        if target_table:
            self.target_table = target_table
            if self._t != target_table._t:
                raise ValueError("Target table must have the same type as the materialized view")
        else:
            # Backward-compatibility path using table_name/engine/order_by_fields
            if not options.table_name:
                raise ValueError("Name of target table is not specified. Provide 'target_table' or 'table_name'.")
            target_table = OlapTable(
                name=options.table_name,
                config=OlapConfig(
                    order_by_fields=options.order_by_fields or [],
                    engine=options.engine
                ),
                t=self._t
            )
        
        if target_table.name == options.materialized_view_name:
            raise ValueError("Target table name cannot be the same as the materialized view name")

        setup = [
            f"CREATE MATERIALIZED VIEW IF NOT EXISTS {quote_identifier(options.materialized_view_name)} TO {quote_identifier(target_table.name)} AS {options.select_statement}",
            f"INSERT INTO {quote_identifier(target_table.name)} {options.select_statement}"
        ]
        teardown = [f"DROP VIEW IF EXISTS {quote_identifier(options.materialized_view_name)}"]

        super().__init__(
            options.materialized_view_name,
            setup,
            teardown,
            pulls_data_from=options.select_tables,
            pushes_data_to=[target_table],
            metadata=options.metadata
        )

        self.target_table = target_table
        self.config = options

File: packages/ts-moose-lib/src/dmv2/sdk/materializedView.ts
import {
  ClickHouseEngines,
  createMaterializedView,
  dropView,
  populateTable,
} from "../../blocks/helpers";
import { Sql, toStaticQuery } from "../../sqlHelpers";
import { OlapTable } from "./olapTable";
import { SqlResource } from "./sqlResource";
import { View } from "./view";
import { IJsonSchemaCollection } from "typia";
import { Column } from "../../dataModels/dataModelTypes";

/**
 * Configuration options for creating a Materialized View.
 * @template T The data type of the records stored in the target table of the materialized view.
 */
export interface MaterializedViewConfig<T> {
  /** The SQL SELECT statement or `Sql` object defining the data to be materialized. Dynamic SQL (with parameters) is not allowed here. */
  selectStatement: string | Sql;
  /** An array of OlapTable or View objects that the `selectStatement` reads from. */
  selectTables: (OlapTable<any> | View)[];

  /** @deprecated See {@link targetTable}
   *  The name for the underlying target OlapTable that stores the materialized data. */
  tableName?: string;
  /** The name for the ClickHouse MATERIALIZED VIEW object itself. */
  materializedViewName: string;

  /** @deprecated See {@link targetTable}
   *  Optional ClickHouse engine for the target table (e.g., ReplacingMergeTree). Defaults to MergeTree. */
  engine?: ClickHouseEngines;

  targetTable?:
    | OlapTable<T> /**  Target table if the OlapTable object is already constructed. */
    | {
        /** The name for the underlying target OlapTable that stores the materialized data. */
        name: string;
        /** Optional ClickHouse engine for the target table (e.g., ReplacingMergeTree). Defaults to MergeTree. */
        engine?: ClickHouseEngines;
      };

  /** Optional ordering fields for the target table. Crucial if using ReplacingMergeTree. */
  orderByFields?: (keyof T & string)[];
}

const requireTargetTableName = (tableName: string | undefined): string => {
  if (typeof tableName === "string") {
    return tableName;
  } else {
    throw new Error("Name of targetTable is not specified.");
  }
};

/**
 * Represents a Materialized View in ClickHouse.
 * This encapsulates both the target OlapTable that stores the data and the MATERIALIZED VIEW definition
 * that populates the table based on inserts into the source tables.
 *
 * @template TargetTable The data type of the records stored in the underlying target OlapTable. The structure of T defines the target table schema.
 */
export class MaterializedView<TargetTable> extends SqlResource {
  /** The target OlapTable instance where the materialized data is stored. */
  targetTable: OlapTable<TargetTable>;

  /**
   * Creates a new MaterializedView instance.
   * Requires the `TargetTable` type parameter to be explicitly provided or inferred,
   * as it's needed to define the schema of the underlying target table.
   *
   * @param options Configuration options for the materialized view.
   */
  constructor(options: MaterializedViewConfig<TargetTable>);

  /** @internal **/
  constructor(
    options: MaterializedViewConfig<TargetTable>,
    targetSchema: IJsonSchemaCollection.IV3_1,
    targetColumns: Column[],
  );
  constructor(
    options: MaterializedViewConfig<TargetTable>,
    targetSchema?: IJsonSchemaCollection.IV3_1,
    targetColumns?: Column[],
  ) {
    let selectStatement = options.selectStatement;
    if (typeof selectStatement !== "string") {
      selectStatement = toStaticQuery(selectStatement);
    }

    if (targetSchema === undefined || targetColumns === undefined) {
      throw new Error(
        "Supply the type param T so that the schema is inserted by the compiler plugin.",
      );
    }

    const targetTable =
      options.targetTable instanceof OlapTable ?
        options.targetTable
      : new OlapTable(
          requireTargetTableName(
            options.targetTable?.name ?? options.tableName,
          ),
          {
            orderByFields: options.orderByFields,
            engine: options.targetTable?.engine ?? options.engine,
          },
          targetSchema,
          targetColumns,
        );
    super(
      options.materializedViewName,
      [
        createMaterializedView({
          name: options.materializedViewName,
          destinationTable: targetTable.name,
          select: selectStatement,
        }),
        populateTable({
          destinationTable: targetTable.name,
          select: selectStatement,
        }),
      ],
      [dropView(options.materializedViewName)],
      {
        pullsDataFrom: options.selectTables,
        pushesDataTo: [targetTable],
      },
    );

    this.targetTable = targetTable;
  }
}
